server/src-lib/Hasura/Eventing/ScheduledTrigger.hs

```diff
@@ -0,0 +1,439 @@
+  where
+    go = unfoldr (fmap dup . nextMatch cron)
+
+-- | Generates events @(from, till]@ according to 'CronSchedule'
+-- generateScheduleTimesFrom :: UTCTime -> UTCTime -> CronSchedule -> [UTCTime]
+-- generateScheduleTimesFrom from till cron = takeWhile (<= till) $ go from
+--   where
+--     go = unfoldr (fmap dup . nextMatch cron)
+
+processScheduledQueue
+  :: HasVersion
+  => L.Logger L.Hasura
+  -> LogEnvHeaders
+  -> HTTP.Manager
+  -> Q.PGPool
+  -> IO SchemaCache
+  -> IO ()
+processScheduledQueue logger logEnv httpMgr pgpool getSC =
+  forever $ do
+    scheduledTriggersInfo <- scScheduledTriggers <$> getSC
+    scheduledEventsE <-
+      runExceptT $
+      Q.runTx pgpool (Q.ReadCommitted, Just Q.ReadWrite) getScheduledEvents
+    case scheduledEventsE of
+      Right partialEvents ->
+        sequence_ $
+        flip map partialEvents $ \(ScheduledEventPartial id' name st payload tries)-> do
+          let sti' = Map.lookup name scheduledTriggersInfo
+          case sti' of
+            Nothing ->  L.unLogger logger $ ScheduledTriggerInternalErr $
+              err500 Unexpected "could not find scheduled trigger in cache"
+            Just sti -> do
+              let webhook = wciCachedValue $ stiWebhookInfo sti
+                  payload' = fromMaybe (fromMaybe J.Null $ stiPayload sti) payload -- override if neccessary
+                  retryConf = stiRetryConf sti
+                  se = ScheduledEventFull id' name st tries webhook payload' retryConf
+              runReaderT (processScheduledEvent logEnv pgpool sti se) (logger, httpMgr)
+      Left err -> L.unLogger logger $ ScheduledTriggerInternalErr $
+        err500 Unexpected $ "could not fetch scheduled events: " <> (T.pack $ show err)
+    threadDelay oneMinute
+
+processScheduledEvent ::
+  ( MonadReader r m
+  , Has HTTP.Manager r
+  , Has (L.Logger L.Hasura) r
+  , HasVersion
+  , MonadIO m
+  )
+  => LogEnvHeaders
+  -> Q.PGPool
+  -> ScheduledTriggerInfo
+  -> ScheduledEventFull
+  -> m ()
+processScheduledEvent logEnv pgpool ScheduledTriggerInfo {..} se@ScheduledEventFull {..} = do
+  currentTime <- liftIO getCurrentTime
+  if diffUTCTime currentTime sefScheduledTime > rcstTolerance stiRetryConf
+    then processDead'
+    else do
+      let timeoutSeconds = round $ rcstTimeoutSec stiRetryConf
+          httpTimeout = HTTP.responseTimeoutMicro (timeoutSeconds * 1000000)
+          headers = map encodeHeader stiHeaders
+          headers' = addDefaultHeaders headers
+          extraLogCtx = ExtraLogContext sefId
+      res <- runExceptT $ tryWebhook headers' httpTimeout sefPayload (T.unpack sefWebhook)
+      logHTTPForST res (Just extraLogCtx)
+      let decodedHeaders = map (decodeHeader logEnv stiHeaders) headers'
+      finally <- either
+                 (processError pgpool se decodedHeaders)
+                 (processSuccess pgpool se decodedHeaders)
+                 res
+      either logQErr return finally
```

Comment:
Could you avoid all these `runExceptT` and `either` calls in this module by programming against `MonadError QErr` rather than returning `Either QErr` explicitly?